ggml: Fix PowerPC build and enable MMA Optimizations

shalinib-ibm opened the pull request:
This change resolves PowerPC build breakage and enables Matrix Math
    Accelerator (MMA) optimizations on supported hardware.
    
    Key changes:
    - Include ggml-cpu PowerPC backend sources in the rsync filter so they are
      propagated from llama/vendor into Ollama’s vendored ggml snapshot.
    - Apply required upstream ggml fixes for PowerPC builds, including:
      - vector macro collision fixes
      - conditional POWER11 backend enablement
    - Enable Matrix Math Accelerator (MMA) support for Power10.
    - Add architecture-specific compiler flags to enable optimized code paths:
      - `-mcpu=power10` when built with the `ppc64le.power10` build tag
        (enables MMA-based kernels, including llamafile_sgemm)
      - `-mcpu=power9` when built with the `ppc64le.power9` build tag
        (enables VSX optimizations)
    
    Build instructions:
    - Power10:
        go build --tags ppc64le.power10 .
    - Power9:
        go build --tags ppc64le.power9 .
    
    Performance impact:
    - ~30% inference time reduction on Power10 with MMA enabled.
    - Measured using:
        ollama run llama3:8b (Q4_0)
        ~50-word summarization, 512-token prompt
      - With MMA: ~6.05s
      - Without MMA: ~8.45s
    
    Improves performance for Q4_0, Q8_0, FP32, and BF16 models on Power10.
    
This change also addresses the PowerPC build breakage noted in the review discussion of #13427.

gabe-l-hart commented on ml/backend/ggml/ggml/src/CMakeLists.txt:
```diff
@@ -403,7 +403,14 @@ if (GGML_CPU_ALL_VARIANTS)
             ggml_add_cpu_backend_variant(power9         POWER9  VSX)
             ggml_add_cpu_backend_variant(power10        POWER10 VSX)
-            ggml_add_cpu_backend_variant(power11        POWER11 VSX)
+	    # POWER11 backend: only if compiler supports -mcpu=power11
```
NIT: indentation
```suggestion
	        # POWER11 backend: only if compiler supports -mcpu=power11
```

gabe-l-hart commented on llama/patches/0033-ggml-conditionally-enable-POWER11-CPU-backend-based-.patch:
```diff
@@ -0,0 +1,36 @@
+             ggml_add_cpu_backend_variant(power9         POWER9  VSX)
+             ggml_add_cpu_backend_variant(power10        POWER10 VSX)
+-            ggml_add_cpu_backend_variant(power11        POWER11 VSX)
++	    # POWER11 backend: only if compiler supports -mcpu=power11
```
NIT: Indentation looks wrong here and that translates to the rendered file as well

gabe-l-hart reviewed (commented):
I'm still a bit confused on the `vector` vs `__vector` macro disambiguation. I don't see anywhere in this PR that defines `__vector`, but I see that `GGML_F32x4` is _always_ mapped to `__vector float` which I think would require that `__vector` always be defined, regardless of the CPU type. Is this guaranteed somewhere?

gabe-l-hart reviewed (commented):
One more indentation NIT (turns out it was a `\t`!), but otherwise I think this should be ready for CI and maintainer review.

@dhiltgen @jmorganca what are the next steps for getting this vetted? I'm guessing you guys don't have a `power` system to test on?

shalinib-ibm commented on llama/patches/0033-ggml-conditionally-enable-POWER11-CPU-backend-based-.patch:
```diff
@@ -0,0 +1,36 @@
+             ggml_add_cpu_backend_variant(power9         POWER9  VSX)
+             ggml_add_cpu_backend_variant(power10        POWER10 VSX)
+-            ggml_add_cpu_backend_variant(power11        POWER11 VSX)
++	           # POWER11 backend: only if compiler supports -mcpu=power11
```
Sorry about this. My bad! Fixed the indent now.

gabe-l-hart reviewed (approved):
I think this looks good by me now!

shalinib-ibm committed:
ggml: Fix PowerPC build and enable MMA optimizations

This change resolves PowerPC build breakage and enables Matrix Math
Accelerator (MMA) optimizations on supported hardware.

Key changes:
- Apply required upstream ggml fixes for PowerPC builds, including:
  - vector macro collision fixes
  - conditional POWER11 backend enablement
- Enable Matrix Math Accelerator (MMA) support for Power10.
- Add architecture-specific compiler flags to enable optimized code paths:
  - `-mcpu=power10` when built with the `ppc64le.power10` build tag
    (enables MMA-based kernels, including llamafile_sgemm)
  - `-mcpu=power9` when built with the `ppc64le.power9` build tag
    (enables VSX optimizations)

Build instructions:
- Power10:
    go build --tags ppc64le.power10 .
- Power9:
    go build --tags ppc64le.power9 .

Performance impact:
- ~30% inference time reduction on Power10 with MMA enabled.
- Measured using:
    ollama run llama3:8b (Q4_0)
    ~50-word summarization, 512-token prompt
  - With MMA: ~6.05s
  - Without MMA: ~8.45s

Improves performance for Q4_0, Q8_0, FP32, and BF16 models on Power10.

shalinib-ibm commented:
> @shalinib-ibm now that #13972 was merged to #13832 and that has been merged to `main`, we should rebuilt this PR to include only the two new patches and the changes they make to the common files.
> 
> @jmorganca Once that's done, this should be a nice tight PR to review that will fully enable `PowerPC` backends!

Thank you so much @gabe-l-hart and @jmorganca for the support on this PR . Have modified this PR to include only the 2 new patch file and corresponding changes in common. files. Kindly review the same.

shalinib-ibm commented:
> Hi , i have try to build the PR on power9 :
> 
> ```
>         .-/+oossssoo+/-.               massimo@ubuntu
>     `:+ssssssssssssssssss+:`           --------------
>   -+ssssssssssssssssssyyssss+-         OS: Ubuntu 22.04.5 LTS ppc64le
> .ossssssssssssssssssdMMMNysssso.       Host: IBM,9009-41A
> ```
> 
> /ssssssssssshdmmNNmmyNMMMMhssssss/ Kernel: 6.8.0-94-generic +ssssssssshmydMMMMMMMNddddyssssssss+ Uptime: 1 day, 19 hours, 8 mins /sssssssshNMMMyhhyyyyhmNMMMNhssssssss/ Packages: 2105 (dpkg), 8 (snap) .ssssssssdMMMNhsssssssssshNMMMdssssssss. Shell: bash 5.1.16 +sssshhhyNMMNyssssssssssssyNMMMysssssss+ Terminal: /dev/pts/0 ossyNMMMNyMMhsssssssssssssshmmmhssssssso CPU: POWER9 (architected) (16) @ 2.3 ossyNMMMNyMMhsssssssssssssshmmmhssssssso Memory: 3017MiB / 24406MiB +sssshhhyNMMNyssssssssssssyNMMMysssssss+ .ssssssssdMMMNhsssssssssshNMMMdssssssss. /sssssssshNMMMyhhyyyyhdNMMMNhssssssss/ +sssssssssdmydMMMMMMMMddddyssssssss+ /ssssssssssshdmNNNNmyNMMMMhssssss/ .ossssssssssssssssssdMMMNysssso. -+sssssssssssssssssyyyssss+- `:+ssssssssssssssssss+:` .-/+oossssoo+/-.
> 
> git clone -b ppc_build_fix https://github.com/shalinib-ibm/ollama.git cd ollama go clean -cache go build --tags ppc64le.power9 .
> 
> but i got this error :
> 
> # github.com/ollama/ollama
> /snap/go/11050/pkg/tool/linux_ppc64le/link: running g++ failed: exit status 1 /usr/bin/g++ -m64 -Wl,--build-id=0xff5d83e084ec67bf68d3945af09abd35a93874e2 -o $WORK/b001/exe/a.out -Wl,--export-dynamic-symbol=_cgo_panic -Wl,--export-dynamic-symbol=_cgo_topofstack -Wl,--export-dynamic-symbol=crosscall2 -Wl,--export-dynamic-symbol=llamaLog -Wl,--export-dynamic-symbol=llamaProgressCallback -Wl,--export-dynamic-symbol=sink -Wl,--compress-debug-sections=zlib /tmp/go-link-4114326907/go.o /tmp/go-link-4114326907/000000.o /tmp/go-link-4114326907/000001.o /tmp/go-link-4114326907/000002.o /tmp/go-link-4114326907/000003.o /tmp/go-link-4114326907/000004.o /tmp/go-link-4114326907/000005.o /tmp/go-link-4114326907/000006.o /tmp/go-link-4114326907/000007.o /tmp/go-link-4114326907/000008.o /tmp/go-link-4114326907/000009.o /tmp/go-link-4114326907/000010.o /tmp/go-link-4114326907/000011.o /tmp/go-link-4114326907/000012.o /tmp/go-link-4114326907/000013.o /tmp/go-link-4114326907/000014.o /tmp/go-link-4114326907/000015.o /tmp/go-link-4114326907/000016.o /tmp/go-link-4114326907/000017.o /tmp/go-link-4114326907/000018.o /tmp/go-link-4114326907/000019.o /tmp/go-link-4114326907/000020.o /tmp/go-link-4114326907/000021.o /tmp/go-link-4114326907/000022.o /tmp/go-link-4114326907/000023.o /tmp/go-link-4114326907/000024.o /tmp/go-link-4114326907/000025.o /tmp/go-link-4114326907/000026.o /tmp/go-link-4114326907/000027.o /tmp/go-link-4114326907/000028.o /tmp/go-link-4114326907/000029.o /tmp/go-link-4114326907/000030.o /tmp/go-link-4114326907/000031.o /tmp/go-link-4114326907/000032.o /tmp/go-link-4114326907/000033.o /tmp/go-link-4114326907/000034.o /tmp/go-link-4114326907/000035.o /tmp/go-link-4114326907/000036.o /tmp/go-link-4114326907/000037.o /tmp/go-link-4114326907/000038.o /tmp/go-link-4114326907/000039.o /tmp/go-link-4114326907/000040.o /tmp/go-link-4114326907/000041.o /tmp/go-link-4114326907/000042.o /tmp/go-link-4114326907/000043.o /tmp/go-link-4114326907/000044.o /tmp/go-link-4114326907/000045.o /tmp/go-link-4114326907/000046.o /tmp/go-link-4114326907/000047.o /tmp/go-link-4114326907/000048.o /tmp/go-link-4114326907/000049.o /tmp/go-link-4114326907/000050.o /tmp/go-link-4114326907/000051.o /tmp/go-link-4114326907/000052.o /tmp/go-link-4114326907/000053.o /tmp/go-link-4114326907/000054.o /tmp/go-link-4114326907/000055.o /tmp/go-link-4114326907/000056.o /tmp/go-link-4114326907/000057.o /tmp/go-link-4114326907/000058.o /tmp/go-link-4114326907/000059.o /tmp/go-link-4114326907/000060.o /tmp/go-link-4114326907/000061.o /tmp/go-link-4114326907/000062.o /tmp/go-link-4114326907/000063.o /tmp/go-link-4114326907/000064.o /tmp/go-link-4114326907/000065.o /tmp/go-link-4114326907/000066.o /tmp/go-link-4114326907/000067.o /tmp/go-link-4114326907/000068.o /tmp/go-link-4114326907/000069.o /tmp/go-link-4114326907/000070.o /tmp/go-link-4114326907/000071.o /tmp/go-link-4114326907/000072.o /tmp/go-link-4114326907/000073.o /tmp/go-link-4114326907/000074.o /tmp/go-link-4114326907/000075.o /tmp/go-link-4114326907/000076.o /tmp/go-link-4114326907/000077.o /tmp/go-link-4114326907/000078.o /tmp/go-link-4114326907/000079.o /tmp/go-link-4114326907/000080.o /tmp/go-link-4114326907/000081.o /tmp/go-link-4114326907/000082.o /tmp/go-link-4114326907/000083.o /tmp/go-link-4114326907/000084.o /tmp/go-link-4114326907/000085.o /tmp/go-link-4114326907/000086.o /tmp/go-link-4114326907/000087.o /tmp/go-link-4114326907/000088.o /tmp/go-link-4114326907/000089.o /tmp/go-link-4114326907/000090.o /tmp/go-link-4114326907/000091.o /tmp/go-link-4114326907/000092.o /tmp/go-link-4114326907/000093.o /tmp/go-link-4114326907/000094.o /tmp/go-link-4114326907/000095.o /tmp/go-link-4114326907/000096.o /tmp/go-link-4114326907/000097.o /tmp/go-link-4114326907/000098.o /tmp/go-link-4114326907/000099.o /tmp/go-link-4114326907/000100.o /tmp/go-link-4114326907/000101.o /tmp/go-link-4114326907/000102.o /tmp/go-link-4114326907/000103.o /tmp/go-link-4114326907/000104.o /tmp/go-link-4114326907/000105.o /tmp/go-link-4114326907/000106.o /tmp/go-link-4114326907/000107.o /tmp/go-link-4114326907/000108.o /tmp/go-link-4114326907/000109.o /tmp/go-link-4114326907/000110.o /tmp/go-link-4114326907/000111.o /tmp/go-link-4114326907/000112.o /tmp/go-link-4114326907/000113.o /tmp/go-link-4114326907/000114.o /tmp/go-link-4114326907/000115.o /tmp/go-link-4114326907/000116.o /tmp/go-link-4114326907/000117.o /tmp/go-link-4114326907/000118.o /tmp/go-link-4114326907/000119.o /tmp/go-link-4114326907/000120.o /tmp/go-link-4114326907/000121.o /tmp/go-link-4114326907/000122.o /tmp/go-link-4114326907/000123.o /tmp/go-link-4114326907/000124.o /tmp/go-link-4114326907/000125.o /tmp/go-link-4114326907/000126.o /tmp/go-link-4114326907/000127.o /tmp/go-link-4114326907/000128.o /tmp/go-link-4114326907/000129.o /tmp/go-link-4114326907/000130.o /tmp/go-link-4114326907/000131.o /tmp/go-link-4114326907/000132.o /tmp/go-link-4114326907/000133.o /tmp/go-link-4114326907/000134.o /tmp/go-link-4114326907/000135.o /tmp/go-link-4114326907/000136.o /tmp/go-link-4114326907/000137.o /tmp/go-link-4114326907/000138.o /tmp/go-link-4114326907/000139.o /tmp/go-link-4114326907/000140.o /tmp/go-link-4114326907/000141.o /tmp/go-link-4114326907/000142.o /tmp/go-link-4114326907/000143.o /tmp/go-link-4114326907/000144.o /tmp/go-link-4114326907/000145.o /tmp/go-link-4114326907/000146.o /tmp/go-link-4114326907/000147.o /tmp/go-link-4114326907/000148.o /tmp/go-link-4114326907/000149.o /tmp/go-link-4114326907/000150.o /tmp/go-link-4114326907/000151.o /tmp/go-link-4114326907/000152.o /tmp/go-link-4114326907/000153.o /tmp/go-link-4114326907/000154.o /tmp/go-link-4114326907/000155.o /tmp/go-link-4114326907/000156.o /tmp/go-link-4114326907/000157.o /tmp/go-link-4114326907/000158.o /tmp/go-link-4114326907/000159.o /tmp/go-link-4114326907/000160.o /tmp/go-link-4114326907/000161.o /tmp/go-link-4114326907/000162.o /tmp/go-link-4114326907/000163.o /tmp/go-link-4114326907/000164.o /tmp/go-link-4114326907/000165.o /tmp/go-link-4114326907/000166.o /tmp/go-link-4114326907/000167.o /tmp/go-link-4114326907/000168.o /tmp/go-link-4114326907/000169.o /tmp/go-link-4114326907/000170.o /tmp/go-link-4114326907/000171.o /tmp/go-link-4114326907/000172.o /tmp/go-link-4114326907/000173.o /tmp/go-link-4114326907/000174.o /tmp/go-link-4114326907/000175.o /tmp/go-link-4114326907/000176.o /tmp/go-link-4114326907/000177.o /tmp/go-link-4114326907/000178.o /tmp/go-link-4114326907/000179.o /tmp/go-link-4114326907/000180.o /tmp/go-link-4114326907/000181.o /tmp/go-link-4114326907/000182.o /tmp/go-link-4114326907/000183.o /tmp/go-link-4114326907/000184.o /tmp/go-link-4114326907/000185.o /tmp/go-link-4114326907/000186.o /tmp/go-link-4114326907/000187.o /tmp/go-link-4114326907/000188.o /tmp/go-link-4114326907/000189.o /tmp/go-link-4114326907/000190.o /tmp/go-link-4114326907/000191.o /tmp/go-link-4114326907/000192.o /tmp/go-link-4114326907/000193.o /tmp/go-link-4114326907/000194.o /tmp/go-link-4114326907/000195.o /tmp/go-link-4114326907/000196.o /tmp/go-link-4114326907/000197.o /tmp/go-link-4114326907/000198.o /tmp/go-link-4114326907/000199.o /tmp/go-link-4114326907/000200.o /tmp/go-link-4114326907/000201.o /tmp/go-link-4114326907/000202.o /tmp/go-link-4114326907/000203.o /tmp/go-link-4114326907/000204.o /tmp/go-link-4114326907/000205.o /tmp/go-link-4114326907/000206.o /tmp/go-link-4114326907/000207.o /tmp/go-link-4114326907/000208.o /tmp/go-link-4114326907/000209.o /tmp/go-link-4114326907/000210.o /tmp/go-link-4114326907/000211.o /tmp/go-link-4114326907/000212.o /tmp/go-link-4114326907/000213.o /tmp/go-link-4114326907/000214.o /tmp/go-link-4114326907/000215.o /tmp/go-link-4114326907/000216.o -O2 -g -lresolv -O2 -g -lpthread -O2 -g -O2 -g -lrt -lpthread -ldl -lstdc++ -lm -O2 -g -O2 -g -O2 -g -O2 -g -O2 -g -O2 -g -O2 -g -O2 -g -O2 -g -no-pie /usr/bin/ld: /tmp/go-link-4114326907/000101.o:(.data.rel.ro+0x48): undefined reference to `ggml_vec_dot_q4_0_q8_0' /usr/bin/ld: /tmp/go-link-4114326907/000101.o:(.data.rel.ro+0x68): undefined reference to `ggml_vec_dot_q4_1_q8_1' /usr/bin/ld: /tmp/go-link-4114326907/000101.o:(.data.rel.ro+0xc8): undefined reference to `ggml_vec_dot_q5_0_q8_0' /usr/bin/ld: /tmp/go-link-4114326907/000101.o:(.data.rel.ro+0xe8): undefined reference to `ggml_vec_dot_q5_1_q8_1' /usr/bin/ld: /tmp/go-link-4114326907/000101.o:(.data.rel.ro+0x100): undefined reference to `quantize_row_q8_0' /usr/bin/ld: /tmp/go-link-4114326907/000101.o:(.data.rel.ro+0x108): undefined reference to `ggml_vec_dot_q8_0_q8_0' /usr/bin/ld: /tmp/go-link-4114326907/000101.o:(.data.rel.ro+0x120): undefined reference to `quantize_row_q8_1' /usr/bin/ld: /tmp/go-link-4114326907/000101.o:(.data.rel.ro+0x148): undefined reference to `ggml_vec_dot_q2_K_q8_K' /usr/bin/ld: /tmp/go-link-4114326907/000101.o:(.data.rel.ro+0x168): undefined reference to `ggml_vec_dot_q3_K_q8_K' /usr/bin/ld: /tmp/go-link-4114326907/000101.o:(.data.rel.ro+0x188): undefined reference to `ggml_vec_dot_q4_K_q8_K' /usr/bin/ld: /tmp/go-link-4114326907/000101.o:(.data.rel.ro+0x1a8): undefined reference to `ggml_vec_dot_q5_K_q8_K' /usr/bin/ld: /tmp/go-link-4114326907/000101.o:(.data.rel.ro+0x1c8): undefined reference to `ggml_vec_dot_q6_K_q8_K' /usr/bin/ld: /tmp/go-link-4114326907/000101.o:(.data.rel.ro+0x208): undefined reference to `ggml_vec_dot_iq2_xxs_q8_K' /usr/bin/ld: /tmp/go-link-4114326907/000101.o:(.data.rel.ro+0x228): undefined reference to `ggml_vec_dot_iq2_xs_q8_K' /usr/bin/ld: /tmp/go-link-4114326907/000101.o:(.data.rel.ro+0x248): undefined reference to `ggml_vec_dot_iq3_xxs_q8_K' /usr/bin/ld: /tmp/go-link-4114326907/000101.o:(.data.rel.ro+0x268): undefined reference to `ggml_vec_dot_iq1_s_q8_K' /usr/bin/ld: /tmp/go-link-4114326907/000101.o:(.data.rel.ro+0x288): undefined reference to `ggml_vec_dot_iq4_nl_q8_0' /usr/bin/ld: /tmp/go-link-4114326907/000101.o:(.data.rel.ro+0x2a8): undefined reference to `ggml_vec_dot_iq3_s_q8_K' /usr/bin/ld: /tmp/go-link-4114326907/000101.o:(.data.rel.ro+0x2c8): undefined reference to `ggml_vec_dot_iq2_s_q8_K' /usr/bin/ld: /tmp/go-link-4114326907/000101.o:(.data.rel.ro+0x2e8): undefined reference to `ggml_vec_dot_iq4_xs_q8_K' /usr/bin/ld: /tmp/go-link-4114326907/000101.o:(.data.rel.ro+0x4e8): undefined reference to `ggml_vec_dot_mxfp4_q8_0' collect2: error: ld returned 1 exit status

Hi @openmax , This fails because ,

The recent update to vendored llama.cpp in which a part of ppc patches went in got reverted.
This is the revert commit: https://github.com/ollama/ollama/pull/14061. 
In order to build on power, please apply the below change also manually along with this PR. 

wget -q -O build_power2.patch https://github.com/ollama/ollama/pull/13972.patch
patch -p1 < build_power2.patch
cmake -B build
cmake --build build
go build --tags ppc64le.power9 -o ollama .
export LD_LIBRARY_PATH=./build/lib/ollama:$LD_LIBRARY_PATH
OLLAMA_NUM_THREADS=<num> ollama serve

This should work now .

shalinib-ibm commented:
CGO engine removed . PR not relevant as MMA optimisations would now come from upstream llama.cpp pinned HEAD.
